### ISSUE TITLE
TypeError when calling super() in decorated method

### ISSUE THREAD

**Issue by kdebrab** [python-attrs/attrs#1038]
Following code:
```python
import attrs


def decorated(method):
    def wrapped(self, *args, **kwargs):
        return method(self, *args, **kwargs)

    return wrapped


@attrs.define()
class A:
    def f(self):
        pass


@attrs.define()
class B(A):
    @decorated
    def f(self):
        super().f()


B().f()
```
raises following error
```
TypeError: super(type, obj): obj must be an instance or subtype of type
```
If I use `super(B, self)`, everything works just fine.

**Comment by hynek** [comment]
This problem is related to cell rewriting for slotted classes. Put simply we have to cheat a bit to make them work by rewriting method contents.

You get the same behavior with dataclasses:

```python
@dataclasses.dataclass(slots=True)
class C:
    @decorated
    def f(self):
        super().f()
```

```
TypeError: super(type, obj): obj must be an instance or subtype of type
```

Short fixes:

- don't use super that way
- disable slots (`@attrs.define(slots=False)`)

Long-term fixes:

- open a bug on CPython for the dataclasses case and we'll steal it from them
- convince @Tinche to fix it for attrs right away ;)

I'm a bit dubious if it's actually possible to fix tho. :-/ **Maybe**, if `functools.wraps` is used and we follow the `__wrapped__` attributes?

**Comment by liblaf** [comment]
I made a few experiments:

**Environment:**

- python: 3.12.10
- attrs: 25.3.0
- wrapt: 1.17.2

<details>

<summary> Same bug as above (click to expand) </summary>

```python
from collections.abc import Callable

import attrs


def decorator[**P, T](func: Callable[P, T]) -> Callable[P, T]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
        print("wrapper()")
        return func(*args, **kwargs)

    return wrapper


@attrs.define
class A:
    def fun(self) -> None:
        print("A.fun()")


@attrs.define
class B(A):
    @decorator
    def fun(self) -> None:
        super().fun()
        print("B.fun()")


b = B()
b.fun()  # TypeError: super(type, obj): obj must be an instance or subtype of type
```

</details>

<details>

<summary> To my surprise, it works. (click to expand) </summary>

```python
from collections.abc import Callable

import attrs


def decorator[**P, T](func: Callable[P, T]) -> Callable[P, T]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
        print("wrapper()")
        return func(*args, **kwargs)

    return wrapper


@attrs.define
class A:
    def fun(self) -> None:
        print("A.fun()")


@attrs.define
class B(A):
    @decorator
    def fun(self) -> None:
        super().fun()
        print("B.fun()")

    def fun_b(self) -> None:
        super().fun()  # ! won't work without this line
        print("B.fun_b()")


b = B()
b.fun()  # It works!
```

</details>

<details>

<summary> wrapt makes everything just fine (click to expand) </summary>

```python
from typing import Any

import attrs
import wrapt


@wrapt.decorator
def decorator(wrapped, instance, args, kwargs) -> Any:
    print("decorator()")
    return wrapped(*args, **kwargs)


@attrs.define
class A:
    def fun(self) -> None:
        print("A.fun()")


@attrs.define
class B(A):
    @decorator
    def fun(self) -> None:
        super().fun()
        print("B.fun()")


b = B()
b.fun()  # It works!
```

</details>

Quoting from [`wrapt`](https://github.com/GrahamDumpleton/wrapt):

> The **wrapt** module focuses very much on correctness. It therefore goes way beyond existing mechanisms such as `functools.wraps()` to ensure that decorators preserve introspectability, signatures, type checking abilities etc. The decorators that can be constructed using this module will work in far more scenarios than typical decorators and provide more predictable and consistent behavior.

I don't know how [`wrapt`](https://github.com/GrahamDumpleton/wrapt) works. Hope this is useful for this issue.

**Comment by hynek** [comment]
Yeah I've been using wrapt a lot too. The frustrating part is that currently we have to choose between runtime correctness (wrapt) and good type hints (functools). At least AFAIK, there's no way to use ParamSpec. :(

**Comment by TrueFurina** [comment]
I hit this same `TypeError` and traced it to `_ClassBuilder._create_slots_class`: when the slotted class is re-created, closure cells referencing the old class are rewritten (see #102), but only for functions directly visible in the class dict. A method wrapped in a decorator hides its function (and its `__class__` cell) behind the wrapper's own closure, so that cell is never rewritten.

A fix is available on `TrueFurina:fix/decorated-super-closure-cell` (commit `5a7e024c`): `_create_slots_class` now also collects functions referenced by other functions' closures before rewriting `__class__` cells. It includes a regression test (`test_decorated_method_no_arg_super` in `tests/test_slots.py`, slots-parametrized). Verified locally against current `main` (8f76777): decorator-wrapped no-arg `super()`, plain methods, class/static methods returning `__class__`, properties, and double-decorator chains all pass. Happy to open a PR if useful.
